### PR TITLE
Write compliant markdown as validated by Markdownlint (issue #3) 

### LINKED ISSUE #3: Keep-a-changelog produces non-compliant Markdown
**Description**
The Markdown produced when saved to a file is not compliant to the Markdown standard and does not pass Markdownlint. This means that when keep-a-changelog is used to edit a compliant file that file becomes non-compliant.

**Example**
Assuming that the keep-a-changelog change log is produced using keep-a-changing (in good dog-fooding tradition) it is an example as this file does not pass Markdownlint.

**Expected behavior**
That the markdown produced passes Markdownlint.

**Actual behavior**
It produces markdown that does not pass Markdownlint. 

**Environment**
 - OS: Linux
 - Rust version: 1.78.0
 - Project version: 0.1.2

**Additional context**
No additional context

### PR DESCRIPTION
**Pull request by gortavoher**
###  Objective 

Ensure that keep-a-changelog writes compliant markdown as validated by Markdownlint. 
Addressing issue #3 from @jerusdp. 
Add a flag permitting a compact output and disabling the header and list lints checking for blank lines.

### Testing 

Implemented data and tests based on following  four scenarios:

- Writing a default changelog as defined by keep-a-changelog defaults
- A default changelog with a single (unreleased) release
- Unreleased with a single Added entry
- Early changelog (based on the current crate ChangeLog) 
- Markdown compliant and compact form examples of each

Each of these cases are tested in two ways:

- create the markdown file using keep-a-changelog. It should be the same as the reference data.
- open the reference data and saved it to another file (read and write back) both should be the same.

### Changes

- implement test data (validated by Markdownlint) and tests for scenarios
- ensure blank lines are printed after each header, final paragraph and final bulleted list item
- remove any residual blank line at the end of the file
- ensure all test cases are passing
- compact flag produces compact form and disables lints checking headers and lists

@napalmpapalam

### CODE REVIEW

File: src/changelog.rs [hunk 1]
```diff
@@ -141,11 +142,26 @@ impl Changelog {
...
+    fn file_contents(&self) -> String {
+        if self.to_string().ends_with("\n\n") {
+            let mut s = self.to_string().trim_end_matches('\n').to_string();
+            s.push('\n');
+            s
+        } else {
+            self.to_string()
+        }
+    }
+
```
Comment (napalmpapalam) [inline src/changelog.rs]:
```suggestion
    /// Format the changelog as a string for output as a valid Markdown file
    ///
    /// To ensure compliance with the requirements of the Markdown standard any blank
    /// line at the end of the string needs to be removed.
    ///
    fn file_contents(&self) -> String {
        if !self.to_string().ends_with("\n\n") {
            return self.to_string();
        }
        
        let mut s = self.to_string().trim_end_matches('\n').to_string();
        s.push('\n');
        s
    }

```

File: tests/data/initial_changelog_unreleased.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [Unreleased]
+
+### Added
+
+- Initial commit
```
Comment (napalmpapalam) [inline tests/data/initial_changelog_unreleased.md:12]: Please, add your changes to the `CHANGELOG.md` file. Nice work btw.

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

**Comment by napalmpapalam** [comment]
Also, could you please make it configurable to add an empty line after the heading 🙏 ?  I understand that it's a best practice, but I stick to not adding it to my projects, coz I like it, it's more readable for me. It could be added by default, but there should be the ability to switch this behavior

**Comment by jerusdp** [comment]
> Also, could you please make it configurable to add an empty line after the heading 🙏 ? I understand that it's a best practice, but I stick to not adding it to my projects, coz I like it, it's more readable for me. It could be added by default, but there should be the ability to switch this behavior

So the default is to keep the markdownlint quite but a feature setting to allow the non-compliant "compact" format.  This will need:
- [ ] Documentation to describe the feature and its impact
- [ ] Tests to validate that it works as expected
- [ ] Feature gated implementation to meet the test. 

As part of this release or as an enhancement?

**Comment by napalmpapalam** [comment]
> > Also, could you please make it configurable to add an empty line after the heading 🙏 ? I understand that it's a best practice, but I stick to not adding it to my projects, coz I like it, it's more readable for me. It could be added by default, but there should be the ability to switch this behavior
> 
> So the default is to keep the markdownlint quite but a feature setting to allow the non-compliant "compact" format. This will need:
> 
> * [ ]  Documentation to describe the feature and its impact
> * [ ]  Tests to validate that it works as expected
> * [ ]  Feature gated implementation to meet the test.
> 
> As part of this release or as an enhancement?

As I see, we can add [here](https://github.com/napalmpapalam/keep-a-changelog-rs/blob/master/src/changelog.rs#L54) one more option, for example, `compact: bool` (`false` by default) and the changelog will be formatted relative to it, everything else seems fine. It could be part of this PR.

**Comment by jerusdp** [comment]
Ok, let me take a crack at it :smiley:

**Comment by jerusdp** [comment]
@napalmpapalam  what is the purpose of the '$' in this code for the writing flags?

```rust
       if let Some(flag) = self.flag.clone() {
            writeln!(f, "<!-- ${flag} -->")?;
        }
```

**Comment by napalmpapalam** [comment]
> @napalmpapalam  what is the purpose of the '$' in this code for the writing flags?
> 
> 
> 
> ```rust
> 
>        if let Some(flag) = self.flag.clone() {
> 
>             writeln!(f, "")?;
> 
>         }
> 
> ```
> 
> 
> 
> 

It's a bug. Plz remove it. I probably missed that thing during rewriting from JS

**Comment by napalmpapalam** [comment]
@gortavoher hey, lemme know (tag me) when u want I will review it, plz

**Comment by jerusdp** [comment]
> @gortavoher hey, lemme know (tag me) when u want I will review it, plz

@napalmpapalam  Good for review now, thanks.

**Review by napalmpapalam — APPROVED** [review]

**Pull request merged**